include/graaflib/algorithm/shortest_path.tpp

```diff
@@ -31,53 +50,29 @@ std::optional<GraphPath<WEIGHT_T>> get_unweighted_shortest_path(
     }
 
     for (const auto& neighbor : graph.get_neighbors(current)) {
-      if (!seen_vertices.contains(neighbor)) {
-        seen_vertices.insert(neighbor);
-        prev_vertex[neighbor] = current;
+      if (!vertex_info.contains(neighbor)) {
+        vertex_info[neighbor] = {
+            neighbor, vertex_info[current].dist_from_start + 1, current};
         to_explore.push(neighbor);
       }
     }
   }
 
-  const auto reconstruct_path = [&start_vertex, &end_vertex, &prev_vertex]() {
-    GraphPath<int> path;
-    auto current{end_vertex};
-
-    while (current != start_vertex) {
-      path.vertices.push_front(current);
-      current = prev_vertex[current];
-    }
-
-    path.vertices.push_front(start_vertex);
-    path.total_weight = path.vertices.size();
-
-    return path;
-  };
-
-  if (seen_vertices.contains(end_vertex)) {
-    return reconstruct_path();
-  } else {
-    return std::nullopt;
-  }
+  return reconstruct_path(start_vertex, end_vertex, vertex_info);
 }
 
 template <typename V, typename E, graph_spec S, typename WEIGHT_T>
 std::optional<GraphPath<WEIGHT_T>> get_weighted_shortest_path(
     const graph<V, E, S>& graph, vertex_id_t start_vertex,
     vertex_id_t end_vertex) {
-  struct DijkstraVertex {
-    vertex_id_t id;
-    WEIGHT_T distance;
-    vertex_id_t previous;
-  };
-
-  std::unordered_map<vertex_id_t, DijkstraVertex> vertex_info;
-  std::priority_queue<
-      DijkstraVertex, std::vector<DijkstraVertex>,
-      std::function<bool(const DijkstraVertex&, const DijkstraVertex&)>>
-      to_explore([](const DijkstraVertex& v1, const DijkstraVertex& v2) {
-        return v1.distance > v2.distance;
-      });
+  std::unordered_map<vertex_id_t, PathVertex<WEIGHT_T>> vertex_info;
+  std::priority_queue<PathVertex<WEIGHT_T>, std::vector<PathVertex<WEIGHT_T>>,
+                      std::function<bool(const PathVertex<WEIGHT_T>&,
+                                         const PathVertex<WEIGHT_T>&)>>
+      to_explore(
+          [](const PathVertex<WEIGHT_T>& v1, const PathVertex<WEIGHT_T>& v2) {
+            return v1.dist_from_start > v2.dist_from_start;
+          });
```

Comment:
What do you think about making this comparison part of the `PathVertex` type? 

```c++
template <typename E>
struct PathVertex {
  // data members

  [[nodiscard]] bool operator>(const PathVertex<E>& other) {
    return dist_from_start > other.dist_from_start;
  }
};
```

Then the type of this priority queue can be simplified to
```suggestion
  using weighted_path_item = PathVertex<WEIGHT_T>;
  std::priority_queue<weighted_path_item, std::vector<weighted_path_item>, std::greater<weighted_path_item>> to_explore{};
```

Reply:
Done! I think you don't need the `weighted_path_item` in `std::greater` though